File: src/n1cspmissingnumber/Java/src/main/com/example/missingnumber/FindMissingNumber.java
// File: FindMissingNumber.java

package src.main.com.example.missingnumber;

public class FindMissingNumber {

    public int findMissingNumber(int[] nums) {
        int lenNums = nums.length;
        int index = 0;

        while (index < lenNums) {
            int value = nums[index];

            if (value < lenNums && value != nums[value]) {
                // Swap nums[index] and nums[value]
                int temp = nums[index];
                nums[index] = nums[value];
                nums[value] = temp;
            } else {
                index++;
            }
        }

        for (int i = 0; i < lenNums; i++) {
            if (i != nums[i]) {
                return i;
            }
        }
        return lenNums;
    }
}

File: src/n1cspmissingnumber/JS/missingnumber/findMissingNumber.js
// File: findMissingNumber.js

function findMissingNumber(nums) {
    let lenNums = nums.length;
    let index = 0;

    // Cyclic Sort
    while (index < lenNums) {
        let value = nums[index];

        if (value < lenNums && value !== nums[value]) {
            [nums[index], nums[value]] = [nums[value], nums[index]];  // Swap
        } else {
            index++;
        }
    }

    // Find the missing number
    for (let i = 0; i < lenNums; i++) {
        if (nums[i] !== i) {
            return i;
        }
    }

    return lenNums;
}

module.exports = findMissingNumber;
